homeassistant/components/yeelight/__init__.py

```diff
@@ -139,73 +152,252 @@
+                        CONF_NIGHTLIGHT_SWITCH, DEFAULT_NIGHTLIGHT_SWITCH
+                    ),
+                },
+            )
 
-    if DOMAIN in config:
-        for ipaddr, device_config in conf[CONF_DEVICES].items():
-            _LOGGER.debug("Adding configured %s", device_config[CONF_NAME])
-            _setup_device(hass, config, ipaddr, device_config)
+        config = {
+            CONF_NAME: entry.data.get(CONF_NAME),  # Support name import from yaml
+            **entry.options,
+        }
+        await _async_setup_device(hass, entry.data[CONF_IP_ADDRESS], config)
+        unset = entry.add_update_listener(_async_update_listener)
+        hass.data[DOMAIN][DATA_CONFIG_ENTRIES][entry.entry_id] = {
+            DATA_UNSUB_UPDATE_LISTENER: unset,
+        }
+        for component in PLATFORMS:
+            hass.async_create_task(
+                hass.config_entries.async_forward_entry_setup(entry, component)
+            )
 
     return True
 
 
-def _setup_device(hass, _, ipaddr, device_config):
-    devices = hass.data[DATA_YEELIGHT]
-
-    if ipaddr in devices:
-        return
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+
+    if unload_ok:
+        if entry.data[CONF_DISCOVERY]:
+            data = hass.data[DOMAIN][DATA_CONFIG_ENTRIES].pop(entry.entry_id)
+            scanner = data[DATA_SCANNER]
+            await scanner.stop_scan()
+            for ipaddr in scanner.seen:
+                hass.data[DOMAIN][DATA_DEVICES].pop(ipaddr)
+            data[DATA_UNSUB_UPDATE_LISTENER]()
+        else:
+            hass.data[DOMAIN][DATA_DEVICES].pop(entry.data[CONF_IP_ADDRESS])
+            data = hass.data[DOMAIN][DATA_CONFIG_ENTRIES].pop(entry.entry_id)
+            data[DATA_UNSUB_UPDATE_LISTENER]()
+
+    return unload_ok
+
+
+async def _async_setup_device(
+    hass: HomeAssistant,
+    ipaddr: str,
+    config: dict,
+    discovery_config_entry: Optional[ConfigEntry] = None,
+) -> None:
+    # Set up device
+    bulb = Bulb(ipaddr, model=config.get(CONF_MODEL) or None)
+    capabilities = await hass.async_add_executor_job(bulb.get_capabilities)
+    if capabilities is None:  # timeout
+        _LOGGER.error("Failed to get capabilities from %s", ipaddr)
+        raise ConfigEntryNotReady
+    device = YeelightDevice(hass, ipaddr, config, bulb)
+    await hass.async_add_executor_job(device.update)
+    hass.data[DOMAIN][DATA_DEVICES][ipaddr] = device
+
+    # Trigger platform setup
+    if discovery_config_entry is not None:
+        data = hass.data[DOMAIN][DATA_CONFIG_ENTRIES][discovery_config_entry.entry_id]
+        hass.async_create_task(data[DATA_SETUP_BINARY_SENSOR](ipaddr))
```

Comment:
Platform discovery entity addition must be signaled with our dispatch helper instead. See eg the ozw integration.